### PR TITLE
[Perpetual]: New Liquidation System

### PR DESCRIPTION
**Pull request by amityadav0**
## Description

- [x] Add new liquidation system similar to leveragelp
- [x] Handle Interest rate in cumulative DS
- [x]  Handle funding rate cumulative DS
- [x] Add closePositions message
- [x] Add migration script
- [x] Add tests

### What has Changed?

What specific problem were you aiming to address, and how did you successfully resolve it? If tests were not uploaded for this pull request or if coverage decreased, please provide an explanation for the change.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/elys-network/elys/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#testing)
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### Deployment Notes

Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.

### Screenshots and Videos

Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.

### CODE REVIEW

File: x/perpetual/keeper/pool.go [hunk 1]
```diff
@@ -68,3 +79,238 @@ func (k Keeper) RemoveLegacyPool(ctx sdk.Context, index uint64) {
...
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.FundingRatePrefix)
+	prev := types.GetFundingRateKey(block-1, pool)
+	key := types.GetFundingRateKey(block, pool)
+	if store.Has(prev) {
+		lastBlock := types.FundingRateBlock{}
+		bz := store.Get(prev)
+		k.cdc.MustUnmarshal(bz, &lastBlock)
+		funding.FundingRate = funding.FundingRate.Add(lastBlock.FundingRate)
+		funding.TotalCustodyLong = funding.TotalCustodyLong.Add(lastBlock.TotalCustodyLong)
+		funding.TotalCustodyShort = funding.TotalCustodyShort.Add(lastBlock.TotalCustodyShort)
```
Comment (cosmic-vagabond) [inline x/perpetual/keeper/pool.go:188]: @amityadav0 why are we adding up block-1 and block numbers, each block should only have knowledge of the funding rate and the long and short custody value as from one block to another you can have an increased on decreased custody amount depending on perpetual activities.

Comment (amityadav0) [reply x/perpetual/keeper/pool.go:188]: this was done similar to interest rate, as interest rate accumulates over time, and if we want to get interest rate between two blocks, we can get that by subtracting previous block
example:
Interest rate 1 1 2 1 1
Store.            1 2 4 5 6
Now to get interest rate between 2-4
we can do position(4) - position(1) = 4

File: x/perpetual/migrations/v6_migration.go
[the remark this answers is not in the document]
Comment (amityadav0) [reply x/perpetual/migrations/v6_migration.go:1]: good catch, thanks
updating

File: x/perpetual/keeper/msg_server_close_positions.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		if err != nil {
+			// Add log about error or not liquidated
+			liqLog = append(liqLog, fmt.Sprintf("Position: Address:%s Id:%d cannot be liquidated due to err: %s", position.Address, position.Id, err.Error()))
+		}
+	}
+
+	// TODO: Handle stop loss
+	ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventClosePositions,
+		sdk.NewAttribute("liquidations", strings.Join(liqLog, "\n")),
+		//sdk.NewAttribute("stop_loss", strings.Join(closeLog, "\n")),
```
Comment (cosmic-vagabond) [inline x/perpetual/keeper/msg_server_close_positions.go:49]: @amityadav0 can we get rid of that?

File: x/perpetual/keeper/keeper.go [hunk 3]
```diff
@@ -504,7 +504,13 @@ func (k Keeper) TakeFundPayment(ctx sdk.Context, returnAmount math.Int, returnAs
...
 	if !takeAmount.IsZero() {
 		takeCoins := sdk.NewCoins(sdk.NewCoin(returnAsset, sdk.NewIntFromBigInt(takeAmount.BigInt())))
-		err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, ammPool.Address, fundAddr, takeCoins)
+
+		ammPoolAddr, err := sdk.AccAddressFromBech32(ammPool.Address)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+		err = k.bankKeeper.SendCoins(ctx, ammPoolAddr, fundAddr, takeCoins)
+		//err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, ammPool.Address, fundAddr, takeCoins)
```
Comment (cosmic-vagabond) [inline x/perpetual/keeper/keeper.go:513]: @amityadav0 can we rid of this comment line?

**Review by cosmic-vagabond — APPROVED** [review]

**Pull request merged**